src/models/geograph.py

```diff
@@ -321,51 +348,36 @@ def _load_from_dataframe(
         Args:
             df (gpd.GeoDataFrame): GeoDataFrame containing polygon objects from
             a shape file.
-            attributes (Optional[List[str]], optional): columns of df that are
-            added to nodes of graph as attributes. Defaults to None,
-            which will cause all attributes to be added.
             tolerance (float, optional): Adds edges between neighbours that are
             at most `tolerance` units apart. Defaults to 0.
 
         Raises:
-            ValueError: If `tolerance` < 0, if `class_label` is not a column in
-            the dataframe, or if `attributes` contains column names not in the
-            dataframe.
+            ValueError: If `tolerance` < 0, if `class_label` or `geometry` are
+            not columns in the dataframe.
 
         Returns:
-            STRtree: The Rtree object used to build the graph.
+            Tuple: A tuple with the Rtree object used to build the graph and the
+            dataframe containing polygon objects.
         """
         if tolerance < 0.0:
             raise ValueError("`tolerance` must be greater than 0.")
-        if (
-            attributes is not None
-            and "class_label" not in attributes
-            or "class_label" not in df.columns
-        ):
+        if "class_label" not in df.columns:
             raise ValueError("`class_label` must be a column in the dataframe.")
-        if (
-            attributes is not None
-            and "geometry" not in attributes
-            or "geometry" not in df.columns
-        ):
+        if "geometry" not in df.columns:
             raise ValueError("`geometry` must be a column in the dataframe.")
-        # If no attribute list is given, all
-        # columns in df are used
-        if attributes is None:
-            attributes = df.columns.tolist()
-        elif not set(attributes).issubset(df.columns):
-            raise ValueError("`attributes` must only contain column names in `df`.")
 
+        # Reset index to ensure consistent indices
+        df = df.reset_index(drop=True)
+        # Using this dict and iterating through it is slightly faster than
+        # iterating through df due to the dataframe overhead
         geom: Dict[int, shapely.Polygon] = df.geometry.to_dict()
-        # Build dict mapping hashable unique object ids for each polygon
-        # to their index in geom
-        id_dict: Dict[int, int] = {
-            id(polygon): index for index, polygon in geom.items()
-        }
+        class_labels: List[int] = df.class_label.tolist()
         # Build Rtree from geometry
-        tree = STRtree(geom.values())
-        # this dict maps polygon indices in df to a list
-        # of neighbouring polygon indices
+        tree = rtree_lib.index.Index()
+        for index, polygon in geom.items():
+            tree.insert(index, polygon.bounds)
```

Comment:
Out of curiosity, this part is super fast right? (From the experience you had so far with respect to the other steps involved in graph creation)
(because when we reproject we probably have to do it as well, and when we merge graphs/add nodes we will also have to update the `rtree`)

Reply:
On a related note, what do you think of factoring out this part of the method as `def _build_rtree(self ...)` for easier readability and in case we want to rebuild `rtrees` in other functions in future?

Reply:
@herbiebradley 
I read up a bit more on this and `geopandas` actually supports a native `rtree` (either in `geos` (written in C) via `pygeos` or with the same `rtree` python library we're using, depending on whether geos is installed).

--> This could be a simple fix to updating and creating the `rtree` and also to transforming it under `crs` operations?
What do you think?


Here's the [gpd code I refer to](https://github.com/geopandas/geopandas/blob/57262a8821ab3c0f348279b6d6fb14512dde8982/geopandas/sindex.py)

Reply:
Doing some tests on this now and it works pretty well, we would simply need to replace/bind `.rtree` with `df.sindex` and use the `.sindex` interface (which also provides a `query_bulk` option which is vectorized if we use geos under the hood).

![image](https://user-images.githubusercontent.com/19754061/109290837-0bf7ae00-7820-11eb-825a-2038dc363e77.png)


It also handles dropping nodes automatically 

![image](https://user-images.githubusercontent.com/19754061/109290900-203bab00-7820-11eb-8c3c-397e94813e7d.png)

Reply:
Yes, this is very fast - and factoring it out seems like a good idea.

Reply:
Update: in 8c0a035 I changed to the spatial index rtree, so factoring out the build_rtree is now unnecessary since the tree gets built whenever you make a `GeoDataFrame`.